### PR TITLE
PyTorch Softmax Ops

### PR DESCRIPTION
**Pull request by HAKSOAT**
## Description

This PR implements the Softmax Ops for PyTorch.

## Related Issue

- [ ] Closes #
- [x] Related to #821 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by HAKSOAT** [comment]
Hi @ricardoV94,

I went through this PR: #764 and I observed that at some point in that work, Softmax, LogSoftmax and SoftmaxGrad were added, in [this commit](https://github.com/HarshvirSandhu/pytensor/blob/629d00bf26416934bfd0c35736efe9be780fb53a/pytensor/link/pytorch/dispatch/elemwise.py#L42-L71).

So I'm curious why it was taken out. I see something about cuda, but was hoping to get more context.

**Comment by ricardoV94** [comment]
> Hi @ricardoV94,
> 
> I went through this PR: #764 and I observed that at some point in that work, Softmax, LogSoftmax and SoftmaxGrad were added, in [this commit](https://github.com/HarshvirSandhu/pytensor/blob/629d00bf26416934bfd0c35736efe9be780fb53a/pytensor/link/pytorch/dispatch/elemwise.py#L42-L71).
> 
> So I'm curious why it was taken out. I see something about cuda, but was hoping to get more context.

We tried to reduce the scope of that initial PR to get the basics in and merge it sooner rather than later. Dropping the implementation of those Ops was probably just that

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 1]
```diff
@@ -34,3 +35,18 @@ def dimshuffle(x):
         return res
 
     return dimshuffle
+
+
+@pytorch_funcify.register(Softmax)
+def pytorch_funcify_Softmax(op, **kwargs):
+    axis = op.axis
+
+    if axis is None:
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:44]: Hmm, None means all axis in PyTensor, doesn't Pytorch support that case?

Comment (HAKSOAT) [reply pytensor/link/pytorch/dispatch/elemwise.py:44]: None in Pytorch implicitly means axis 1: https://discuss.pytorch.org/t/implicit-dimension-choice-for-softmax-warning/12314/10

And that funcationality is deprecated now, giving the warning:

> UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  torch.nn.functional.softmax(torch.Tensor(x), dim=None)

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py:44]: So can we write the function to achieve the same behavior as softmax along all axis? I guess in that case we need to do it manually with `exp(x) / torch.sum(exp(x), axis=None)` or is that also not allowed?

Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/elemwise.py:44]: Yes, reshaping can be used for `axis=None`
Also, it might be a good idea to make sure `x` is a `torch.float` dtype. The softmax would fail if the dtype is int or long.

Comment (HAKSOAT) [reply pytensor/link/pytorch/dispatch/elemwise.py:44]: @ricardoV94 , what is the expected behaviour when the input is of dtype int?

I'm thinking of converting to float implicitly.

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py]: Have to check the integer case but don't forget to test it as well

Comment (HAKSOAT) [reply pytensor/link/pytorch/dispatch/elemwise.py]: Hi @ricardoV94, I have added the Logsoftmax and Softmaxgrad ops.

I have a question in relation to testing the types. How do I do this?

Everytime I send the functiongraph and the inputs into `compare_pytorch_and_py`, the function always ends up with a Tensor of floats somehow.

File: environment.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ channels:
 dependencies:
   - python>=3.10
   - compilers
-  - numpy>=1.17.0
+  - numpy>=1.17.0,<2
```
Comment (HAKSOAT) [inline environment.yml:12]: This change is because of this error: #827

File: pytensor/link/pytorch/dispatch/elemwise.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py]: In the test you're creating a `matrix` which has a default dtype of `float32` or `float64`, you can parametrize the test with an explicit dtype, say `pytest.mark.parametrize("dtype", ["float64", "int64"])` to test integer types as well.

This is something we should mention in the docs @HarshvirSandhu

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 3]
```diff
@@ -34,3 +35,46 @@ def dimshuffle(x):
...
     return dimshuffle
+
+
+@pytorch_funcify.register(Softmax)
+def pytorch_funcify_Softmax(op, **kwargs):
+    axis = op.axis
+
+    def softmax(x):
+        if not torch.is_floating_point(x):
+            x = x.to(torch.float32)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:46]: Should probably convert to the output type advertised by PyTensor. The second argument of the funcify function, is `node` (now hidden inside **kwargs. If you retrieve `node`, you can check the dtype of the output via `node.outputs[0].dtype`. This is probably what you should convert to, not necessarily float32
```suggestion
            x = x.to(torch.float32)
```

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py:46]: Actually there's a bug in PyTensor Softmax, it also fails if you try to execute with dtype. Can you open a separate issue for that in the repo here? For now it's enough for the torch dispatch function to raise a `NotImplementedError` if the output dtype is an integer, and not try to handle it inside the returned `softmax` function.

Same for the other Softmax related functions

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/elemwise.py:46]: Opened an issue: https://github.com/pymc-devs/pytensor/issues/857

File: pytensor/link/pytorch/dispatch/elemwise.py [hunk 4]
```diff
@@ -40,11 +40,14 @@ def dimshuffle(x):
 @pytorch_funcify.register(Softmax)
 def pytorch_funcify_Softmax(op, **kwargs):
     axis = op.axis
+    dtype = kwargs["node"].outputs[0].dtype
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/elemwise.py:43]: Maybe better to check the input dtype? Sorry if I said the output before
```suggestion
    dtype = kwargs["node"].inputs[0].dtype
```

**Comment by ricardoV94** [comment]
Left one comment, after that I think it's ready!

**Comment by HAKSOAT** [comment]
Hi @ricardoV94 , I have made that change and resolved existing merge conflicts.

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Thanks @HAKSOAT
